src/test/groovy/au/org/ala/userdetails/PasswordServiceSpec.groovy

```diff
@@ -0,0 +1,335 @@
+package au.org.ala.userdetails
+
+import grails.testing.gorm.DataTest
+import grails.testing.services.ServiceUnitTest
+import org.passay.RuleResultMetadata
+
+class PasswordServiceSpec extends UserDetailsSpec implements ServiceUnitTest<PasswordService>, DataTest {
+
+    def charLength = RuleResultMetadata.CountCategory.Length
+    def charLowerCase = RuleResultMetadata.CountCategory.LowerCase
+    def charUpperCase = RuleResultMetadata.CountCategory.UpperCase
+    def charDigit = RuleResultMetadata.CountCategory.Digit
+    def charSpecial = RuleResultMetadata.CountCategory.Special
+    def charWhitespace = RuleResultMetadata.CountCategory.Whitespace
+    def charAllowed = RuleResultMetadata.CountCategory.Allowed
+    def charIllegal = RuleResultMetadata.CountCategory.Illegal
+
+    void setupSpec() {
+        mockDomains(Role, User, Password)
+    }
+
+    def setup() {
+        service.passwordEncoderType = 'bcrypt'
+        service.legacyAlgorithm = null
+        service.legacySalt = null
+    }
+
+    def cleanup() {
+        User.deleteAll()
+        Role.deleteAll()
+        UserRole.deleteAll()
+        UserProperty.deleteAll()
+    }
+
+    void 'test service passwordEncoderType is as expected'() {
+        expect:
+        service.passwordEncoderType == PasswordService.BCRYPT_ENCODER_TYPE
+    }
+
+    void 'test reset password generates and stores a new password for the user'() {
+        given:
+        def tempAuthKey = "wzAdTZYn1xJDJrjwHgpu"
+        def user = createUser(tempAuthKey)
+        def newPassword = "0hFAnO9dWq6rcUopZ9EN"
+
+        when:
+        service.resetPassword(user, newPassword)
+
+        then:
+        0 * _ // no other interactions
+        Password.count() == 1
+        service.comparePasswords(newPassword, Password.first().password)
+        Password.first().user == user
+    }
+
+    void 'test generate new password generates and stores a new password for the user'() {
```

Comment:
Password resets should not be modifying the users password, otherwise malicious users can reset anyones password and lock them out temporarily.

Reply:
This service method is only used by the admin only screen that generates a user a new password and emails it to them.